src/rapid_pro_to_engagement_db/rapid_pro_to_engagement_db.py

```diff
@@ -1,12 +1,58 @@
+from datetime import timedelta
+
 from core_data_modules.cleaners import URNCleaner
 from core_data_modules.logging import Logger
 from engagement_database.data_models import Message, MessageDirections, MessageStatuses, HistoryEntryOrigin
-from rapid_pro_tools.rapid_pro_client import RapidProClient
-from storage.google_cloud import google_cloud_utils
+
+from src.rapid_pro_to_engagement_db.cache import RapidProSyncCache
 
 log = Logger(__name__)
 
 
+def _get_contacts_from_cache(cache=None):
+    """
+    :param cache: Cache to check for contacts. If None, returns None.
+    :type cache: src.rapid_pro_to_engagement_db.cache.RapidProSyncCache | None
+    :return: Contacts from a cache, if the cache exists and a previous contacts file exists in the cache, else None.
+    :rtype: list of temba_client.v2.Contact | None
+    """
+    if cache is None:
+        return None
+    else:
+        return cache.get_contacts()
+
+
+def _get_new_runs(rapid_pro, flow_id, flow_result_field, cache=None):
+    """
+    Gets new runs from Rapid Pro for the given flow.
+
+    If a cache is provided and it contains a timestamp of a previous export, only returns runs that have been modified
+    since the last export.
+
+    :param rapid_pro: Rapid Pro client to use to download new runs.
+    :type rapid_pro: rapid_pro_tools.rapid_pro_client.RapidProClient
+    :param flow_id: Flow id to download runs for.
+    :type flow_id: str
+    :param flow_result_field: Result field in the flow.
+    :type flow_result_field: str
+    :param cache: Cache to check for a timestamp of a previous export. If None, downloads all runs.
+    :type cache: src.rapid_pro_to_engagement_db.cache.RapidProSyncCache | None
+    :return: Runs modified for the given flow since the cache was last updated, if possible, else from all of time.
+    :rtype: list of temba_client.v2.Run
+    """
+    # Try to get the last modified timestamp from the cache
+    flow_last_updated = None
+    if cache is not None:
+        flow_last_updated = cache.get_latest_run_timestamp(flow_id, flow_result_field)
+
+    # If there is a last updated timestamp in the cache, only download and return runs that have been modified since.
+    filter_last_modified_after = None
+    if flow_last_updated is not None:
+        filter_last_modified_after = flow_last_updated + timedelta(microseconds=1)
```

Comment:
we can use timedelta.resolution, but feel free to leave as is.

Reply:
Thanks, I've checked the docs for .resolution and still think this is clearer because (i) we don't need to look up what the resolution is, (ii) this needs to be textit's resolution rather than python's, which are the same by coincidence.